// ts2c/src/standard/global/isNaN.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var template_1 = require("../../template");
var standard_1 = require("../../standard");
var ctypes_1 = require("../../types/ctypes");
var typeconvert_1 = require("../../nodes/typeconvert");
var IsNaNResolver = /** @class */ (function () {
    function IsNaNResolver() {
    }
    IsNaNResolver.prototype.matchesNode = function (typeHelper, call) {
        return call.expression.kind === ts.SyntaxKind.Identifier && call.expression.getText() === "isNaN";
    };
    IsNaNResolver.prototype.argumentTypes = function (typeHelper, call) {
        return [ctypes_1.UniversalVarType];
    };
    IsNaNResolver.prototype.returnType = function (typeHelper, call) {
        return ctypes_1.BooleanVarType;
    };
    IsNaNResolver.prototype.createTemplate = function (scope, node) {
        return new CIsNaN(scope, node);
    };
    IsNaNResolver.prototype.needsDisposal = function (typeHelper, node) {
        return false;
    };
    IsNaNResolver.prototype.getTempVarName = function (typeHelper, node) {
        return null;
    };
    IsNaNResolver.prototype.getEscapeNode = function (typeHelper, node) {
        return null;
    };
    IsNaNResolver = __decorate([
        standard_1.StandardCallResolver
    ], IsNaNResolver);
    return IsNaNResolver;
}());
var CIsNaN = /** @class */ (function (_super) {
    __extends(CIsNaN, _super);
    function CIsNaN(scope, call) {
        var _this = _super.call(this) || this;
        _this.argument = null;
        _this.argument = new typeconvert_1.CAsUniversalVar(scope, call.arguments[0]);
        scope.root.headerFlags.js_var_isnan = true;
        scope.root.headerFlags.js_var_to_number = true;
        return _this;
    }
    CIsNaN = __decorate([
        template_1.CodeTemplate("js_var_isnan({argument})")
    ], CIsNaN);
    return CIsNaN;
}(template_1.CTemplateBase));
