// ts2c/src/standard/string/charCodeAt.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var template_1 = require("../../template");
var standard_1 = require("../../standard");
var ctypes_1 = require("../../types/ctypes");
var elementaccess_1 = require("../../nodes/elementaccess");
var StringCharCodeAtResolver = /** @class */ (function () {
    function StringCharCodeAtResolver() {
    }
    StringCharCodeAtResolver.prototype.matchesNode = function (typeHelper, call) {
        if (call.expression.kind != ts.SyntaxKind.PropertyAccessExpression)
            return false;
        var propAccess = call.expression;
        var objType = typeHelper.getCType(propAccess.expression);
        return propAccess.name.getText() == "charCodeAt" && objType == ctypes_1.StringVarType;
    };
    StringCharCodeAtResolver.prototype.objectType = function (typeHelper, call) {
        return ctypes_1.StringVarType;
    };
    StringCharCodeAtResolver.prototype.argumentTypes = function (typeHelper, call) {
        return call.arguments.map(function (a, i) { return i == 0 ? ctypes_1.NumberVarType : null; });
    };
    StringCharCodeAtResolver.prototype.returnType = function (typeHelper, call) {
        return ctypes_1.NumberVarType;
    };
    StringCharCodeAtResolver.prototype.createTemplate = function (scope, node) {
        return new CStringSearch(scope, node);
    };
    StringCharCodeAtResolver.prototype.needsDisposal = function (typeHelper, node) {
        return false;
    };
    StringCharCodeAtResolver.prototype.getTempVarName = function (typeHelper, node) {
        return null;
    };
    StringCharCodeAtResolver.prototype.getEscapeNode = function (typeHelper, node) {
        return null;
    };
    StringCharCodeAtResolver = __decorate([
        standard_1.StandardCallResolver
    ], StringCharCodeAtResolver);
    return StringCharCodeAtResolver;
}());
var CStringSearch = /** @class */ (function (_super) {
    __extends(CStringSearch, _super);
    function CStringSearch(scope, call) {
        var _this = _super.call(this) || this;
        var propAccess = call.expression;
        _this.topExpressionOfStatement = call.parent.kind == ts.SyntaxKind.ExpressionStatement;
        if (!_this.topExpressionOfStatement) {
            if (call.arguments.length == 1) {
                _this.strAccess = new elementaccess_1.CElementAccess(scope, propAccess.expression);
                _this.position = template_1.CodeTemplateFactory.createForNode(scope, call.arguments[0]);
                scope.root.headerFlags.str_char_code_at = true;
            }
            else
                console.log("Unsupported number of parameters in " + call.getText() + ". Expected one parameter.");
        }
        return _this;
    }
    CStringSearch = __decorate([
        template_1.CodeTemplate("\n{#if !topExpressionOfStatement}\n    str_char_code_at({strAccess}, {position})\n{/if}")
    ], CStringSearch);
    return CStringSearch;
}(template_1.CTemplateBase));
