// ts2c/ts2c.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var program_1 = require("./src/program");
function transpile(source) {
    var sourceFile = ts.createSourceFile('source.ts', source, ts.ScriptTarget.ES5, true);
    var compilerHost = {
        getSourceFile: function (fileName, target) { return 'source.ts' ? sourceFile : null; },
        writeFile: function (name, text, writeByteOrderMark) { },
        getDefaultLibFileName: function () { return "lib.d.ts"; },
        useCaseSensitiveFileNames: function () { return false; },
        getCanonicalFileName: function (fileName) { return fileName; },
        getCurrentDirectory: function () { return ""; },
        getDirectories: function () { return []; },
        getNewLine: function () { return "\n"; },
        fileExists: function (fileName) { return fileName == 'source.ts'; },
        readFile: function (fileName) { return fileName == 'source.ts' ? source : null; },
        directoryExists: function (dirName) { return dirName == ""; },
    };
    var program = ts.createProgram(['source.ts'], { noLib: true }, compilerHost);
    return new program_1.CProgram(program)["resolve"]();
}
exports.transpile = transpile;
;
