// ts2c/src/standard/number/number.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var template_1 = require("../../template");
var standard_1 = require("../../standard");
var ctypes_1 = require("../../types/ctypes");
var NumberCallResolver = /** @class */ (function () {
    function NumberCallResolver() {
    }
    NumberCallResolver.prototype.matchesNode = function (typeHelper, call) {
        return ts.isIdentifier(call.expression) && call.expression.text == "Number";
    };
    NumberCallResolver.prototype.returnType = function (typeHelper, call) {
        var type = typeHelper.getCType(call.arguments[0]);
        return type == ctypes_1.NumberVarType ? ctypes_1.NumberVarType : ctypes_1.UniversalVarType;
    };
    NumberCallResolver.prototype.createTemplate = function (scope, node) {
        return new CNumberCall(scope, node);
    };
    NumberCallResolver.prototype.needsDisposal = function (typeHelper, node) {
        return false;
    };
    NumberCallResolver.prototype.getTempVarName = function (typeHelper, node) {
        return null;
    };
    NumberCallResolver.prototype.getEscapeNode = function (typeHelper, node) {
        return null;
    };
    NumberCallResolver = __decorate([
        standard_1.StandardCallResolver
    ], NumberCallResolver);
    return NumberCallResolver;
}());
var CNumberCall = /** @class */ (function (_super) {
    __extends(CNumberCall, _super);
    function CNumberCall(scope, call) {
        var _this = _super.call(this) || this;
        _this.call = "";
        _this.arg = template_1.CodeTemplateFactory.createForNode(scope, call.arguments[0]);
        var type = scope.root.typeHelper.getCType(call.arguments[0]);
        if (type != ctypes_1.NumberVarType && type != ctypes_1.UniversalVarType) {
            _this.call = "str_to_int16_t";
            scope.root.headerFlags.str_to_int16_t = true;
        }
        else if (type == ctypes_1.UniversalVarType) {
            _this.call = "js_var_to_number";
            scope.root.headerFlags.js_var_to_number = true;
        }
        return _this;
    }
    CNumberCall = __decorate([
        template_1.CodeTemplate("\n{#if call}\n    {call}({arg})\n{#else}\n    {arg}\n{/if}")
    ], CNumberCall);
    return CNumberCall;
}(template_1.CTemplateBase));
